src/plugins/intel_cpu/src/shape_inference/custom/subgraph.hpp

```diff
@@ -10,34 +10,91 @@ namespace ov {
 namespace intel_cpu {
 namespace node {
 using Result = IShapeInfer::Result;
-/* This class implementation is a temporal WA
-   TODO: revise the implementation to remove the node reference*/
 class SnippetShapeInfer : public ShapeInferEmptyPads {
 public:
-    SnippetShapeInfer(Snippet* node) : m_node(node) {}
+    SnippetShapeInfer(std::shared_ptr<ov::Model> body) : m_body(body) {}
```

Comment:
In general case, subgraph's `validate_nodes_and_infer_types` is actually used: https://github.com/openvinotoolkit/openvino/pull/15430/files#diff-0cedaaf60bc2518ef9628e775185635df0ddc5f4b1fad1a7de1ca6aa7e875320R65,
Just in case of specific(one output), A quicker one is implemented. Unified validate_nodes_and_infer_types should be used if it's optimized in the future.

Reply:
I want to avoid Snippets users to implement custom code for shape inference themselves. 
We have PR that adopts new memory accessors for shape inference: https://github.com/openvinotoolkit/openvino/pull/17719. That should significantly improve shape inference via IStaticShapeInfer interface. 
If custom impl still will be required it cab be implemented reusing existing approach https://github.com/openvinotoolkit/openvino/tree/master/src/core/shape_inference/include, but placed into snippets component.

Reply:
@dmitry-gorokhov, I agree that shape inference implementation should be hidden inside op::Subgraph. However we still need a wrapper of some kind to bridge cpu and snippets shape_inference, but it should be as simple as forwarding arguments: subgraph->shape_infer(....), or smth like that.
Anyway, snippets shape-inference will be refactored and moved to op::Subgraph in [PR 18563](https://github.com/openvinotoolkit/openvino/pull/18563), so I would be happy to address your concern there.